src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -208,25 +230,46 @@ private void DiscoverTestsPrivate(ITestDiscoveryEventsHandler discoveryEventsHan
         /// Each concurrent discoverer calls this method, once its completed working on previous data
         /// </summary>
         /// <param name="ProxyDiscoveryManager">Proxy discovery manager instance.</param>
-        /// <returns>True, if discovery triggered</returns>
-        private bool DiscoverTestsOnConcurrentManager(IProxyDiscoveryManager proxyDiscoveryManager)
+        private void DiscoverTestsOnConcurrentManager(IProxyDiscoveryManager proxyDiscoveryManager)
         {
-            DiscoveryCriteria discoveryCriteria = null;
-
-            string nextSource = null;
-            if (this.TryFetchNextSource(this.sourceEnumerator, out nextSource))
+            // Peek to see if we have sources to trigger a discovery
+            if (this.TryFetchNextSource(this.sourceEnumerator, out string nextSource))
             {
-                EqtTrace.Info("ProxyParallelDiscoveryManager: Triggering test discovery for next source: {0}", nextSource);
-                discoveryCriteria = new DiscoveryCriteria(new List<string>() { nextSource }, this.actualDiscoveryCriteria.FrequencyOfDiscoveredTestsEvent, this.actualDiscoveryCriteria.DiscoveredTestEventTimeout, this.actualDiscoveryCriteria.RunSettings);
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("ProxyParallelDiscoveryManager: Triggering test discovery for next source: {0}", nextSource);
+                }
+
+                // Kick off another discovery task for the next source
+                var discoveryCriteria = new DiscoveryCriteria(new[] { nextSource }, this.actualDiscoveryCriteria.FrequencyOfDiscoveredTestsEvent, this.actualDiscoveryCriteria.DiscoveredTestEventTimeout, this.actualDiscoveryCriteria.RunSettings);
+                Task.Run(() =>
+                    {
+                        if (EqtTrace.IsVerboseEnabled)
+                        {
+                            EqtTrace.Verbose("ParallelProxyDiscoveryManager: Discovery started.");
+                        }
+
+                        proxyDiscoveryManager.DiscoverTests(discoveryCriteria, this.GetHandlerForGivenManager(proxyDiscoveryManager));
+                    })
+                    .ContinueWith(t =>
+                    {
+                        // Just in case, the actual discovery couldn't start for an instance. Ensure that
+                        // we call discovery complete since we have already fetched a source. Otherwise
+                        // discovery will not terminate
+                        if (EqtTrace.IsWarningEnabled)
+                        {
+                            EqtTrace.Warning("ParallelProxyDiscoveryManager: Failed to trigger discovery. Exception: " + t.Exception);
+                        }
+
+                        this.GetHandlerForGivenManager(proxyDiscoveryManager).HandleDiscoveryComplete(0, Enumerable.Empty<TestCase>(), true);
```

Comment:
On abort, totalTests need to be -1, otherwise it breaks assert [here](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs#L21).